agent/lib_aws_sdk_php.c

```diff
@@ -519,6 +518,171 @@ char* nr_lib_aws_sdk_php_get_command_arg_value(char* command_arg_name,
   return command_arg_value;
 }
 
+void nr_lib_aws_sdk_php_dynamodb_set_params(
+    nr_segment_datastore_params_t* datastore_params,
+    nr_segment_cloud_attrs_t* cloud_attrs,
+    NR_EXECUTE_PROTO) {
+  zval* endpoint_zval = NULL;
+  zval* region_zval = NULL;
+  zval* host_zval = NULL;
+  zval* port_zval = NULL;
+  zval* this_obj = NULL;
+  zend_function* func = NULL;
+  zend_class_entry* base_class = NULL;
+  char* table_name = NULL;
+  char* account_id = NULL;
+
+  if (NULL == datastore_params || NULL == cloud_attrs) {
+    return;
+  }
+
+  this_obj = NR_PHP_USER_FN_THIS();
+  func = nr_php_execute_function(NR_EXECUTE_ORIG_ARGS);
+
+  if (NULL == this_obj || NULL == func) {
+    return;
+  }
+
+  if (NULL != func->common.scope) {
+    base_class = func->common.scope;
+
+    region_zval = nr_php_get_zval_object_property_with_class(
+        this_obj, base_class, "region");
+    if (nr_php_is_zval_non_empty_string(region_zval)) {
+      cloud_attrs->cloud_region = Z_STRVAL_P(region_zval);
+    }
+
+    endpoint_zval = nr_php_get_zval_object_property_with_class(
+        this_obj, base_class, "endpoint");
+    if (nr_php_is_zval_valid_object(endpoint_zval)) {
+      host_zval = nr_php_get_zval_object_property(endpoint_zval, "host");
+      if (nr_php_is_zval_non_empty_string(host_zval)) {
+        datastore_params->instance->host = Z_STRVAL_P(host_zval);
+
+        /* Only try to get a port if we have a valid host. */
+        port_zval = nr_php_get_zval_object_property(endpoint_zval, "port");
+        if (nr_php_is_zval_valid_integer(port_zval)) {
+          /* Must be freed by caller */
+          datastore_params->instance->port_path_or_id
+              = nr_formatf(NR_INT64_FMT, Z_LVAL_P(port_zval));
+        } else {
+          /* In case where host was found but port was not, spec says return
+           * unknown for port. */
+          datastore_params->instance->port_path_or_id = nr_strdup("unknown");
+        }
+      }
+    }
+  }
+  if (NULL == datastore_params->instance->host) {
+    /* Unable to retrieve the endpoint, go with AWS defaults. */
+    datastore_params->instance->host = AWS_SDK_PHP_DYNAMODBCLIENT_DEFAULT_HOST;
+    /* Need to strdup because the calling function will free it. */
+    datastore_params->instance->port_path_or_id
+        = nr_strdup(AWS_SDK_PHP_DYNAMODBCLIENT_DEFAULT_PORT);
+  }
+
+  table_name = nr_lib_aws_sdk_php_get_command_arg_value(
+      AWS_SDK_PHP_DYNAMODBCLIENT_TABLENAME_ARG, NR_EXECUTE_ORIG_ARGS);
+  if (!nr_strempty(table_name)) {
+    /* Must be freed by caller */
+    datastore_params->collection = table_name;
+  }
+  if (!nr_strempty(NRINI(aws_account_id))) {
+    account_id = NRINI(aws_account_id);
+  }
+
+  if (NULL != datastore_params->collection && NULL != account_id
+      && NULL != cloud_attrs->cloud_region) {
+    /* Must be freed by caller */
+    cloud_attrs->cloud_resource_id = nr_formatf(
```

Comment:
What would be the best behavior here if `cloud_attrs->cloud_resource_id` was non-NULL?

Reply:
Best behavior?  if non-NULL it gets passed back to caller who uses it to end the segment before freeing it.

Reply:
I am curious if this is an expected situation so it needs to be handled to not leak memory, or if it is not expected then what is the best course?

Reply:
It is expected to be non-null for dynamodb but regardless of null/non-null, the calling function [knows](https://github.com/newrelic/newrelic-php-agent/pull/1037/files#diff-b5410abe33a951704c8600c7b0b214d3c431cee94dcb3089567ec0ec9f328fc1R656) and will [call nr_free](https://github.com/newrelic/newrelic-php-agent/pull/1037/files#diff-b5410abe33a951704c8600c7b0b214d3c431cee94dcb3089567ec0ec9f328fc1R681) which will either free or ignore if NULL.